generators/chipyard/src/main/scala/CustomBusTopologies.scala

```diff
@@ -0,0 +1,100 @@
+
+package chipyard
+
+import freechips.rocketchip.config.{Field, Config, Parameters}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util.{Location, Symmetric}
+import freechips.rocketchip.subsystem._
+
+// I'm putting this code here temporarily as I think it should be a candidate
+// for upstreaming based on input from Henry Cook, but don't wnat to deal with
+// an RC branch just yet.
+
+// For subsystem/BusTopology.scala
+
+/**
+  * Keys that serve as a means to define crossing types from a Parameters instance
+  */
+case object SubsystemCrossingParamsKey extends Field[SubsystemCrossingParams](SubsystemCrossingParams())
+case object MemoryBusCrossingTypeKey extends Field[ClockCrossingType](NoCrossing)
+
+// Biancolin: This, modified from Henry's email
+/** Parameterization of a topology containing a banked coherence manager and a bus for attaching memory devices. */
+case class CoherentBusTopologyParams(
+  sbus: SystemBusParams, // TODO remove this after better width propagation
+  mbus: MemoryBusParams,
+  l2: BankedL2Params,
+  sbusToMbusXType: ClockCrossingType = NoCrossing
+) extends TLBusWrapperTopology(
+  instantiations = (if (l2.nBanks == 0) Nil else List(
+    (MBUS, mbus),
+    (L2, CoherenceManagerWrapperParams(mbus.blockBytes, mbus.beatBytes, l2.nBanks, L2.name)(l2.coherenceManager)))),
+  connections = if (l2.nBanks == 0) Nil else List(
+    (SBUS, L2,   TLBusWrapperConnection(xType = NoCrossing, driveClockFromMaster = Some(true), nodeBinding = BIND_STAR)()),
+    (L2,  MBUS,  TLBusWrapperConnection.crossTo(
+      xType = sbusToMbusXType,
+      driveClockFromMaster = Some(true),
```

Comment:
No, since the slave-side clock is bound to a different diplomatic source in the master based on the crossing type.

Reply:
So both connections are made in the case where the clocks are different, and we rely on last connect semantics? Or is it somehow explicitly overridden in the diplomacy code.

I trust that its correct and you don't need to change anything but I'm just curious.

Reply:
My understanding is that there's no reliance on chisel last connect semantics. A single connection between master wrapper and the slave wrapper is generated (because driveClockFromMaster = true). The diplomatic origin of the clock in the master defines what provides the RHS, and that is determined based on the crossing type provided as the first argument in `TLBusWrapperConnection` methods.

Specifically, that happens here: 
https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/tilelink/BusWrapper.scala#L181

If we're using an async crossing, the slave bus (MBUS) will show up as its own clock group in the AsyncClockGroups in our Subsystem. Anything else and it shows up as another clock in the master bus's clock group (this is why there are so many "sbus"-prefixed clocks by default -- in a single-clock system everything is eventually coupled to the sbus's clock group).